### PR TITLE
Feature/media mime fetch

### PR DESCRIPTION
**Pull request by nbradbury**
As mentioned in [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/6172#issuecomment-311694550) on #6172, in order to support pagination with the new media filters, we first need to update FluxC to support the `mime_type` parameter when fetching media, then update WPAndroid to fetch the specific mime type when the user changes the filter.

~~This PR is on hold until [the related FluxC branch](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/491) has been merged.~~

cc: @daniloercoli

### CODE REVIEW

**Comment by daniloercoli** [comment]
It seems that "load more" feature doesn't work until you manually refresh the list.
See the video https://cloudup.com/cVkzMcsizHR

**Comment by nbradbury** [comment]
> It seems that "load more" feature doesn't work until you manually refresh the list.

Sorry, looks like I broke that in a recent commit. Fixed in c2d26a3.

**Comment by daniloercoli** [comment]
Cool, load more seems to work fine now, but I'm back with some issues:

- Empty placeholder view can be blank:
-- Go to documents, refresh it. The empty view says that I don't have documents. 
-- Switch to Videos
-- Switch back to documents. Empty view is blank.
device-2017-06-30-143749

- Switching tab starts the refresh. See the video https://cloudup.com/cIxLgQ9ecFU

- Videos section refreshing acting weirdly https://cloudup.com/cIxLgQ9ecFU

File: WordPress/src/main/java/org/wordpress/android/ui/media/MediaGridFragment.java [hunk 1]
```diff
@@ -508,7 +541,22 @@ private void fetchMediaList(boolean loadMore) {
             updateEmptyView(EmptyViewMessageType.LOADING);
             setRefreshing(true);
 
-            FetchMediaListPayload payload = new FetchMediaListPayload(mSite, loadMore);
+            // use a partial mime_type match when fetching media
+            String mimeType;
+            switch (mFilter) {
```
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaGridFragment.java]: I noticed that the network requests aren't including the mime type - the reason is that the cases in this switch are missing `break`s.

**Comment by nbradbury** [comment]
> Cool, load more seems to work fine now, but I'm back with some issues:

@daniloercoli I was pulling my hair out over these, but then I discovered that FluxC [clears the media store](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java#L721) after every fetch. I'll work on a FluxC PR that clears only the media matching the requested mime type, which should correct these.

**Comment by nbradbury** [comment]
@daniloercoli This should all be working now, thanks to @aforcier's quick review of my related [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/497).

> Switching tab starts the refresh. 

That's actually on purpose. The first time you visit a tab a fetch is performed to make sure we have all the media items matching that filter. Perhaps it would be good if I performed the fetch but didn't show the swipe-to-refresh spinner if the grid is already populated from the local store?

**Comment by nbradbury** [comment]
> I had troubles updating media on a site of mine and found out that handleFetchAllMediaError is missing from setRefreshing(false)

Fixed in 86123fe.

> Should we also check wether the error is about the current filter?

That makes sense, but we don't know what the mime type is from the error. Knowing that would require adding a `mimeType` to the FluxC [MediaError](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java#L158), but `MediaError` is used in several places so I don't think we should change it.

**Comment by daniloercoli** [comment]
> Knowing that would require adding a mimeType to the FluxC MediaError, but MediaError is used in several places so I don't think we should change it

Yeah, you're right. This is probably one of those kind of situations where having the ability to set a tag, or having an ID for each single request, will help to fix the issue without changing FluxC.

Can't we just extends `MediaError`, add the mime type info to the new error claz, and change the `fetchMediaList` [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java#L245) ? cc @aforcier

**Comment by nbradbury** [comment]
> Can't we just extends MediaError, add the mime type info to the new error claz, and change the fetchMediaList here ?

That would work. I'll wait for @aforcier to chime in before making that change.

**Comment by aforcier** [comment]
> Can't we just extends MediaError, add the mime type info to the new error claz, and change the fetchMediaList here ? cc @aforcier

`OnMediaListFetched` already returns the `mimeType` - could we instead [pass that through](https://github.com/wordpress-mobile/WordPress-Android/blob/e1bedd136cceab5db0192c77ba67a747fed01ea2/WordPress/src/main/java/org/wordpress/android/ui/media/MediaGridFragment.java#L346) to `handleFetchAllMediaError()`?

**Comment by nbradbury** [comment]
> OnMediaListFetched already returns the mimeType - could we instead pass that through to handleFetchAllMediaError()?

bdbfa3577583492294b406d5ed1a006c

Added in 76096ea :)

**Comment by daniloercoli** [comment]
> OnMediaListFetched already returns the mimeType - could we instead pass that through to handleFetchAllMediaError()?

@aforcier Are you sure of this? It seems to me that `mimeType` is not populated in case of errors.  https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java#L724

**Comment by aforcier** [comment]
> @aforcier Are you sure of this? It seems to me that mimeType is not populated in case of errors. https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java#L724

Yes, you're right - we should be populating the `mimeType` field in the `FetchMediaListResponsePayload` from the network clients even on error, and passing it through to the `OnMediaListFetched` event in the store (though we don't need to change the `MediaError`).

**Comment by daniloercoli** [comment]
I've created a PR for that here https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/502

**Comment by nbradbury** [comment]
@daniloercoli @aforcier Thanks for that FluxC PR and merge, I'll update this PR shortly.

**Comment by nbradbury** [comment]
@daniloercoli I updated this PR to use the latest FluxC hash and verified that we're getting the mime type in `handleFetchAllMediaError()`, so ready for another (dare I hope last?) review.

**Comment by daniloercoli** [comment]
It seems to me that `load more` doesn't work anymore. Tested on Android 5.

**Comment by nbradbury** [comment]
> It seems to me that load more doesn't work anymore.

Looks like I failed to push my last commit, which updated the FluxC hash. I just did that and tested on Android 5 without any problems. Can you give it another try?

**Comment by daniloercoli** [comment]
Looks good! (Even though I noticed that load more doesn't show the refreshing spinner anymore. Maybe it is on purpose. If not we can fix this in the next PR).

:shipit:

**Pull request merged**